promql/parser/prettier.go

```diff
@@ -0,0 +1,162 @@
+// 2. Does the current Node needs to be prettified?
+//
+// The level of a Node determines if it should be indented or not.
+// The answer to the 1 is NO if the level passed is 0. This means, the
+// parent Node did not apply a new line, so the current Node must not
+// apply any indentation as prefix.
+// If level > 1, a new line is applied by the parent. So, the current Node
+// should prefix an indentation before writing any of its content. This indentation
+// will be ([level/depth of current Node] * "  ").
+//
+// The answer to 2 is YES if the normalized length of the current Node exceeds
+// the maxCharactersPerLine limit. Hence, it applies the indentation equal to
+// its depth and increments the level by 1 before passing down the child.
+// If the answer is NO, the current Node returns the normalized string value of itself.
+
+var maxCharactersPerLine = 100
+
+func Prettify(n Node) string {
+	return n.Pretty(0)
+}
+
+func (e *AggregateExpr) Pretty(level int) string {
+	s := indent(level)
+	if !needsSplit(e) {
+		s += e.String()
+		return s
+	}
+
+	s += e.getAggStr()
+	s += "(\n"
+
+	if e.Op.IsAggregatorWithParam() {
+		s += fmt.Sprintf("%s,\n", e.Param.Pretty(level+1))
+	}
+	s += fmt.Sprintf("%s\n%s)", e.Expr.Pretty(level+1), indent(level))
+	return s
+}
+
+func (e *BinaryExpr) Pretty(level int) string {
+	s := indent(level)
+	if !needsSplit(e) {
+		s += e.String()
+		return s
+	}
+	returnBool := ""
+	if e.ReturnBool {
+		returnBool = " bool"
+	}
+
+	matching := e.getMatchingStr()
+	return fmt.Sprintf("%s\n%s%s%s%s\n%s", e.LHS.Pretty(level+1), indent(level), e.Op, returnBool, matching, e.RHS.Pretty(level+1))
+}
+
+func (e *Call) Pretty(level int) string {
+	s := indent(level)
+	if !needsSplit(e) || len(e.Args) == 1 {
+		s += e.String()
+		return s
+	}
+	s += fmt.Sprintf("%s(\n%s\n%s)", e.Func.Name, e.Args.Pretty(level+1), indent(level))
+	return s
+}
+
+func (e *EvalStmt) Pretty(_ int) string {
+	return "EVAL " + e.Expr.String()
+}
+
+func (e Expressions) Pretty(level int) string {
+	if !needsSplit(e) {
+		return e.String()
+	}
+	s := ""
+	for i := range e {
+		s += fmt.Sprintf("%s,\n", e[i].Pretty(level))
+	}
+	return s[:len(s)-2]
+}
+
+func (e *ParenExpr) Pretty(level int) string {
+	s := indent(level)
+	if !needsSplit(e) {
+		s += e.String()
+		return s
+	}
+	return fmt.Sprintf("%s(\n%s\n%s)", s, e.Expr.Pretty(level+1), indent(level))
+}
+
+func (e *MatrixSelector) Pretty(level int) string {
+	return getCommonPrefixIndent(level, e)
+}
+
+func (e *SubqueryExpr) Pretty(level int) string {
+	return getCommonPrefixIndent(level, e)
+}
+
+func (e *VectorSelector) Pretty(level int) string {
+	return getCommonPrefixIndent(level, e)
+}
+
+func (e *NumberLiteral) Pretty(level int) string {
+	return getCommonPrefixIndent(level, e)
+}
+
+func (e *StringLiteral) Pretty(level int) string {
+	return getCommonPrefixIndent(level, e)
+}
+
+func (e *UnaryExpr) Pretty(level int) string {
+	return getCommonPrefixIndent(level, e)
+}
```

Comment:
Can the unary expression be long? Like `-histogram_quantile(0.99, sum by (le) (rate(...)))`

Reply:
Applying formatting to unary is tricky. This is because there can be 2 cases
1. Input: `-histogram_quantile(0.99, sum by (le) (rate(...)))`
Output:
```
-histogram_quantile(
  0.9,
  rate(foo[1m])
)
```
2. Input: `(-histogram_quantile(0.99, sum by (le) (rate(...))))`
Output:
```
-  histogram_quantile(
    0.9,
    rate(foo[1m])
  )
```
This is because we have to increment the level by 1, so the child (histogram_quantile) applied a prefix indent. Technically, we should pass down level as 0 so that the child does not apply prefix indent. But the moment we pass down level as 0, the above expression will be

```
(
  -histogram_quantile(
  0.9,
  rate(foo[1m])
)
)
```

Hence, I think, we should not split UnaryExpr

Reply:
histogram_quantile functions are pretty common, and can be long, so we should split actually.

To think of it in a simple way, here is how I see this to be prettified:

For `-histogram_quantile(0.99, sum by (le) (rate(...)))`, first prettify the function without `-`

```
histogram_quantile(
  0.9,
  rate(foo[1m])
)
```

And then replace the `function_name` with `- function_name`

```
- histogram_quantile(
  0.9,
  rate(foo[1m])
)
```

How about this?

Reply:
Maybe that does not look that pretty, but it's a start until we have better ways defined.

Reply:
Actually this suggestion of yours looks better
```
-histogram_quantile(
  0.9,
  rate(foo[1m])
)
```

Reply:
I think we have to go with replace with function name method. The suggestion in the above comment will cause problems since it resets the level to 0.